crates/evm/src/execute.rs

```diff
@@ -2,12 +2,73 @@
 
 use reth_primitives::{BlockNumber, BlockWithSenders, Receipt, Receipts, Request, Requests, U256};
 use reth_prune_types::PruneModes;
-use revm::db::BundleState;
-use revm_primitives::db::Database;
+use revm::{db::BundleState, Context, DatabaseCommit, Evm};
+use revm_primitives::{db::Database, EVMError, ResultAndState};
 
 pub use reth_execution_errors::{BlockExecutionError, BlockValidationError};
 pub use reth_storage_errors::provider::ProviderError;
 
+/// An EVM executor trait that executes an input (e.g. transaction and height) and produces an output
+/// (e.g. state changes).
+///
+/// The executor cannot commit the state changes to the database directly, see [`EvmCommitter`].
+pub trait EvmExecutor {
+    /// The environment for the executor.
+    type Env;
+    /// The output produced by the executor.
+    type Output;
+    /// The error produced by the executor.
+    type Error;
```

Comment:
using associated types we can add proper restrictions so that these can be `filled` with header + tx, I think that makes sense